### PR TITLE
Separate InstrumentationLibrary from metric.Descriptor

### LINKED ISSUE #2086: Metrics: create independent MeterProvider
The existing Metrics SDK supports a MeterProvider API but the underlying components treat the Resource as part of the export Record and do not formally recognize a division between Meters.   This issue outlines a proposal to solve this. The outcome will be that each:

1. metric.Descriptor no longer includes instrumentation name and version, nor Resource
2. metric/registry supports a UniqueMeterProvider that maintains a map of named MeterImpls.
3. metrics Export() includes additional SourceData (instrumentation name and version, schema URL, Resource)
4. sdk/metric/controller maintains once accumulator/checkpointer per Meter, exports them separately
5. OTLP exporter for metrics no longer needs to group data, is much simpler.

This is too much change for a single PR, but I have prototyped it in https://github.com/open-telemetry/opentelemetry-go/pull/2085. This issue will cover breaking up the draft PR into smaller pieces.

Another goal is met in the draft PR above, that metric.Descriptor could be moved out of the top-level metric package into a sub-package (possibly named "sdkapi"), as demonstrated in https://github.com/open-telemetry/opentelemetry-go/pull/2044.

### PR DESCRIPTION
**Pull request by jmacd**
Please excuse a relatively large PR. I've made efforts to keep it easy-to-review as described below.

The problem being addressed is that the `metric.Descriptor` contains the instrumentation library name and version, which force consumers of the metric data to re-group metrics by instrumentation library.

One goal is to move the Descriptor type into the new `metric/sdkapi` sub-package, however that move will only create a larger PR so this PR leaves it in place and a future, simple PR will apply the move. We had to refactor this because the `InstrumentOption` and `MeterOption` types do not belong in `sdkapi`. Thus, `NewDescriptor` needed to lose those options.

The `metric.NewDescriptor()` function was problematic because it took both `InstrumentOption` and `MeterOption`, and our goal is to disentangle these. A few valid non-test uses of `NewDescriptor` are modified to manage the `InstrumentOption` values themselves, i.e., compute an InstrumentConfig and extract the two options (description and unit) manually. All the test-only uses of `metric.NewDescriptor()` are replaced by `metrictest.NewDescriptor()` with the old behavior (i.e., accepting varargs options for description and unit).  This minimizes the size of this PR.

The metric Controller now maintains one Accumulator per instrumentation library, as opposed to one Accumulator per SDK. This changes the export path, giving it a two-level structure where the exporter sees one instrumentation library at a time, then can visit each export record.

The name `CheckpointSet` became problematic and was never a good name. "Checkpoint" is the write-oriented behavior that is carried out by the metric Processor; "Reader" is the read-oriented behavior carried out by the exporter. The existing `CheckpointSet` has been renamed `Reader`. A new type named `InstrumentationLibraryReader` is introduced to read one library at a time, which gives a two-level `ForEach()` style of iteration. 

The Controller now takes a factory for building processors instead of a single processor, thus many constructors change from `controller.New(processor.New(...), ...)` to `controller.New(processor.NewFactory(...), ...)`.

The real, functional changes in this PR are:

1. in `metric/registry` a new `*Provider` type maintains a unique-map of `Meter` implementations, separate from the unique-map of instrument/kind that it already kept.  The new `*Provider` wraps around an implementation of `MeterProvider` that is not required to maintain name-to-Meter uniqueness, whereas before this was encapsulated in a single unique map.
2. in `internal/metric/global` a new `*Provider` type defers MeterProvider initialization explicitly, whereas before the uniqueness was implicitly managed by the registry
3. in `sdk/metric/controller/basic`, now using the registry Provider, which exposes a `List()` method to access the registered instrumentation libraries, to export two-level information through the new `InstrumentationLibraryReader`. there is a slight change of lock semantics; whereas before the Controller used its one Accumulator's state as a lock to determine when refresh was needed, it now uses the same lock used to protect against multiple Start()/Stop() calls; no significant change of behavior here
4. in `exporters/otlp/otlpmetric/...` the grouping by instrumentation library is removed; calls to `source()` and `sink()` are carried out once per instrumentation library.
5. the metric SDK properly implements the instrumentation library SchemaURL field (and is tested)

A bit of test cleanup has been applied. There were a number of test-only implementations of the former `CheckpointSet` interface that were consolidated into helpers in `sdk/metric/processor/processortest`.

Part of #2090.
Fixes #2086.
Part of #2119.

File-by-file review notes (non-test files): https://github.com/open-telemetry/opentelemetry-go/pull/2197#issuecomment-919559499

### CODE REVIEW

**Comment by jmacd** [comment]
@evantorrie please take a look?

**Review by codeboten — COMMENTED** [review]
Still working through the PR, just wanted to share the feedback so far

File: exporters/otlp/otlpmetric/internal/otlpmetrictest/otlptest.go
Comment (jmacd) [reply exporters/otlp/otlpmetric/internal/otlpmetrictest/otlptest.go:43]: that would ... explain a few mysteries.

File: sdk/resource/resource.go
Comment (jmacd) [reply sdk/resource/resource.go:136]: https://github.com/open-telemetry/opentelemetry-go/pull/2227

Comment (jmacd) [reply sdk/resource/resource.go:136]: I've removed these lines here and fixed the test to avoid a nil resource.

Comment (jmacd) [reply sdk/resource/resource.go]: 65b8f3e0ea91fa7fdbdce8caf2311e89f0dc40e1

File: exporters/otlp/otlpmetric/exporter.go
Comment (jmacd) [reply exporters/otlp/otlpmetric/exporter.go:59]: https://github.com/open-telemetry/opentelemetry-go/issues/2228

**Comment by jmacd** [comment]
@open-telemetry/go-approvers This PR is causing trouble for others. @tigrannajaryan has spent time adding a flag that is also covered by this PR. I'm sorry to have to apply such large refactorings, but if we allow only small PRs and never a large PR this SDK will never come together. Please approve #2197.

**Review by codeboten — APPROVED** [review]

**Comment by jmacd** [comment]
`bridge/opencensus/exporter.go`

This code is injecting OpenCensus metric data into an OTel export pipeline.  Whereas `CheckpointSet.ForEach` iterated over one list of records, the pipeline now supports iteration over list of records per instrumentation library.  This code simply injects the library name "OpenCensus bridge" as the library, whereas before it was set in every `*metric.Descriptor`.

`exporters/otlp/otlpmetric/exporter.go`

A call to `CheckpointSet()` becomes `InstrumentationLibraryReader()` and a TODO about the apperance of a singleton slice that could be eliminated later.

`exporters/otlp/otlpmetric/internal/metrictransform/metric.go`

`CheckpointSet()` is renamed `InstrumentationLibraryReader()`.  The body of the former method (i.e., `source()` through `sink()`) is indented one level inside the top-level `ForEach()`, constructing one library of metric data at a time.  Inside `source()` there is a single-level map instead of the former two-level map.

`exporters/prometheus/prometheus.go`

A call to `c.exp.Controller().ForEach()` with a function literal becomes a two-level iteration with two function literals.  The inner body doesn't change.  This happens twice, once for the Prometheus `Describe()` and once for `Collect()`, but is just re-indenting.

`exporters/stdout/stdoutmetric/metric.go`

A pattern is developing: replace a one-level iteration with a two-level iteration, code otherwise unchanged.

`internal/metric/global/meter.go`

The global meter has to remember the instrumentation library Name, Version and SchemaURL and pass the name and library options to the delegate constructor.

`metric/config.go`

`InstrumentConfig` loses two fields (instrumentation library name and version).  `MeterConfig` gains one field (schema URL). `InstrumentMeterOption` is gone.  `WithInstrumentationVersion()` is now just a `MeterOption`.  `WithSchemaURL()` is a new `MeterOption` to bring this code into parity with the tracing SDK.

`metric/metric.go`

`NewDescriptor()` no longer refers to instrument options, takes just the structural options `Description` and `Unit`.  This method will move into `sdkapi` in the future.  Non-test uses of this method are renamed to use test packages to avoid adding new explicit arguments to this method.

`metric/metric_sdkapi.go`

The `Meter` object no longer knows its instrumentation name and version, that knowledge enters much later in the export pipeline now.

`metric/registry/registry.go`

This class implements the logic needed by any SDK (including global) to maintain a unique map of Meters.  This was formerly not an explicit concept because the underlying implementation maintained a flat list of instruments, each of which knew its library.  Provides a simple getter to iterate over all meters.

`sdk/export/metric/metric.go`

A one-level iteration becomes a two-level iteration.  `CheckpointSet` becomes the outer `InstrumentationLibraryReader`, and the new inner `Reader` interface is added.  `Reader` maintains the `sync.Locker` portion of the interface.  Independent libraries are now locked independently for read/write on the export path, to allow greater potential read/write concurrency.

`sdk/metric/controller/basic/controller.go`

These are real changes.

The controller changes to allocate one accumulator per library, as a basis for the new two-level iteration.  `type accumulatorProvider` implements `metric.MeterProvider` and is called by the new `registry.MeterProvider` to construct new Meter implementations. `type accumulatorCheckpointer` is the record of a single library, where it stores the meter details (i.e., `instrumentation.Library`), the accumulator, and the checkpointer.

New method `func (c *Controller) Reader()` exposes the `InstrumentationLibraryReader` for pull-based exporters to use.  `type libraryReader` implements the top-level iteration over the registry's list of meters.  The lower-level iteration is the same as before.

Minor: the `shouldCollect()` method uses the Controller's lock instead of the Accumulator's lock, to determine whether the next export interval has been reached by the export loop.

`sdk/metric/processor/basic/basic.go`

A new `NewFactory()` method makes it easy for the Controller (and tests) to repeatedly construct new controllers.

File: sdk/metric/controller/basic/pull_test.go
Comment (jmacd) [reply sdk/metric/controller/basic/pull_test.go:54]: Fixed 😁

File: bridge/opencensus/go.mod
Comment (MrAlias) [inline bridge/opencensus/go.mod:11]:
```suggestion
	go.opentelemetry.io/otel/sdk/metric v0.23.0
```

right?

File: metric/registry/registry.go
Comment (MrAlias) [inline metric/registry/registry.go:68]: formatting nit:

```suggestion
		m = metric.WrapMeterImpl(NewMeterImpl(
			p.impl.Meter(instrumentationName, opts...).MeterImpl(),
		))
```

File: metric/registry/registry.go
Comment (MrAlias) [inline metric/registry/registry.go:141]: Why are we able to lessen the restriction here and only check if the name is not already registered instead of the full descriptor?

File: sdk/metric/controller/basic/controller.go
Comment (MrAlias) [inline sdk/metric/controller/basic/controller.go:165]:
```suggestion
// InstrumentationLibraryReader returns an InstrumentationLibraryReader for iterating
// through the metrics of each registered library, one at a time.
func (c *Controller) InstrumentationLibraryReader() export.InstrumentationLibraryReader {
```

File: sdk/metric/controller/basic/controller.go
Comment (MrAlias) [inline sdk/metric/controller/basic/controller.go:261]: Can we rename this to better describe the function it performs.

File: sdk/metric/controller/basic/controller.go
Comment (MrAlias) [inline sdk/metric/controller/basic/controller.go:254]: Do we want this type assertion to panic?

File: sdk/metric/controller/basic/controller.go
Comment (MrAlias) [inline sdk/metric/controller/basic/controller.go:316]: Similar here, should this panic if the type assertion fails? Do we want to skip the impl and log an error instead? Is there a way we could design this so the type assertion isn't necessary?

File: sdk/metric/controller/controllertest/test.go
Comment (MrAlias) [inline sdk/metric/controller/controllertest/test.go:71]:
```suggestion
	return reader.ForEach(func(library instrumentation.Library, reader export.Reader) error {
		return reader.ForEach(kind, func(record export.Record) error {
```

File: metric/registry/registry.go
Comment (jmacd) [reply metric/registry/registry.go:141]: The former code had a 1-level map by (instrument name, library name, version). 

There are now two levels of mapping, the `MeterProvider` maintains a unique mapping of a 3-part key (instrumentation name, version, schema_url) to Meter. This level of the mapping ensures all Meters share underlying state.

This object, adapted from the prior code, maintains a 1-part key which is just instrument name. At this level in the code, the mapping ensures all instruments are registered with the same type.

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go]: Yes, I should've done this in the first place!

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go:254]: The panic would only result from a bug. The provider is a registry owned by this object that will call this code to produce the corresponding value, so it didn't seem necessary to test the error condition.

I've added an "impossible type assertion failed: %T" to to defend against the impossible (in two places) 😁

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go:261]: Changed to 
```
// checkpointSingleAccmulator checkpoints a single instrumentation
// library's accumulator, which involves calling
// checkpointer.StartCollection, accumulator.Collect, and
// checkpointer.FinishCollection in sequence.
func (c *Controller) checkpointSingleAccmulator(ctx context.Context, ac *accumulatorCheckpointer) error {
```

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go:316]: Fixed with "impossible type assertion failure" error.

> Is there a way we could design this so the type assertion isn't necessary?

Yes, but it would lead to duplication of the registry code, which is used by the global and by the real metric controller. If we had generics, the registry would use them and this would not require a type assertion.

If we copy the registry code into at least two places, there wouldn't be a type assertion, but that doesn't fee like a win. The registry code is responsible for ensuring that duplicate instruments are registered with consistent types, so both the real and global implementations are supposed to do this.

**Review by MrAlias — APPROVED** [review]
LGTM outside of fixing the build failure (looks like some renaming needs to be propagated).

:rocket:

**Comment by jmacd** [comment]
I look forward to making fixes in go-contrib 😁

File: metric/registry/registry.go
Comment (evantorrie) [inline metric/registry/registry.go:89]: s/NewMeter/NewMeterImpl/

Comment (jmacd) [reply metric/registry/registry.go]: Fixed!

File: sdk/metric/controller/basic/controller.go
Comment (evantorrie) [inline sdk/metric/controller/basic/controller.go:269]: s/checkpointSingleAccmulator/checkpointSingleAccumulator/

File: sdk/metric/processor/basic/basic.go
Comment (evantorrie) [inline sdk/metric/processor/basic/basic.go:116]: At line 93 (not shown here), comment still refers to `state.RWMutex` implementing locking for the `CheckpointSet` interface. I believe this should be changed to `Reader` interface.

File: exporters/otlp/otlpmetric/internal/metrictransform/metric.go
Comment (evantorrie) [inline exporters/otlp/otlpmetric/internal/metrictransform/metric.go:135]: preexisting typo s/encoutered/encountered/

**Review by evantorrie — APPROVED** [review]
Looks good to me in general. Mostly just comments around some remaining references to `CheckpointSet` (e.g. check out `sdk/metric/doc.go`).

File: metric/config.go
Comment (jmacd) [reply metric/config.go]: True. I hadn't noticed. I've changed two methods to match the other style, 52b6ca95

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go:345]: I've added a TODO about this and will file a new issue before we merge this.

File: metric/registry/registry.go
Comment (jmacd) [reply metric/registry/registry.go]: After your feedback I studied this code in greater detail and realized more simplifications:

1. the metric/registry package was entirely internal, so moved it to internal/metric/registry
2. registry.MeterProvider was only being used in Controller, so I've moved it there, but this also requires an explanation:
2a. the global package did not use the registry MeterProvider because it has specific synchronization requirements and because it's also required to keep track of the meter details to perform its function. Note the global package still uses the unique instrument registry function, so there are still two users of the code in internal/metric/registry
2b. the metrictest package doesn't use the registry, "just because" the tests don't include uniqueness checking for meters or instruments.

As for `sync.Map`, I don't think there's a great need for concurrency. This lock would prevent concurrent instrument registration from collection activities, but I don't expect that in practice. To use the `sync.Map` requires a bit more work because you have to construct an object for `LoadAndStore()`, and since this object is trying to establish at-most-once semantics, it would require another lock somewhere.

Moving the MeterProvider functionality into the Controller simplified this PR a bit further. Please take another look specifically at the new implementation in controller.go.

File: metric/registry/registry.go
Comment (jmacd) [reply metric/registry/registry.go]: Thanks. This code is rolled into the controller now.

File: metric/registry/registry.go
Comment (jmacd) [reply metric/registry/registry.go]: Since this code moved into the controller, now it's using the controller's lock. This required a slight change in `Shutdown()` to release the lock before the final collection.

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go]: Do you think we still need methods like `InstrumentationLibraryReader()` and `MeterProvider()` given that they just return `c`, which directly implements the other interfaces?

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go]: I've rewritten this section completely -- the Controller implements its own provider logic.

The registry still provides a MeterImpl, used in two places, that does instrument uniqueness checking. The registry is now an internal package, so I added a method to "poke through" the abstraction and get the underlying `MeterImpl`, which allows the controller to get back to its `*accumulatorCheckpointer`.

File: exporters/otlp/otlpmetric/internal/metrictransform/metric.go
Comment (jmacd) [reply exporters/otlp/otlpmetric/internal/metrictransform/metric.go]: This word was misspelled a bunch in this one file -- fixed.

File: metric/registry/registry.go
Comment (jmacd) [reply metric/registry/registry.go]: I moved this logic into the Controller and fixed some of the other comments nearby in this file. This file is now located in `internal/`.

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go:115]: Fixed.

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go]: This was kind of awful and it ended up a lot nicer after a recent change. The controller implements its own MeterProvider logic; the registry.UniqueInstrumentMeterImpl provides a method to "poke through" to the underlying implementation, which the controller uses to get back to its own object.

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go]: Fixed!

File: sdk/metric/processor/basic/basic.go
Comment (jmacd) [reply sdk/metric/processor/basic/basic.go:116]: I've replaced CheckpointSet with Reader everywhere now.

**Comment by evantorrie** [comment]
LGTM.

File: sdk/metric/controller/basic/controller.go
Comment (MadVikingGod) [reply sdk/metric/controller/basic/controller.go]: If they are needed to satisfy some other interface then please remove them.  The smaller the API area the better.

File: metric/registry/registry.go
Comment (MadVikingGod) [reply metric/registry/registry.go]: It's fine without the sync.Map.  The reason I asked was that Maps have a built-in at most one with [Map.Load**Or**Store](https://pkg.go.dev/sync#Map.LoadOrStore).  I get there is overhead because of the translation via interfaces, maybe this kind of change would be better suited when we can use generics.

**Review by MadVikingGod — APPROVED** [review]
I'm happy with where it is, if you find you can remove the functions discussed above then :+1: to that too.

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go]: Done.

File: metric/registry/registry.go
Comment (jmacd) [reply metric/registry/registry.go]: I was imagining the contract of the registry is that it will not call e.g. NewSyncInstrument() on the underlying implementation more than once. The aggregation pipeline expects a unique `*MetricDescriptor` for its purposes, so to use `LoadOrStore()` here I imagine an intermediate object, with a pattern like:

```
  newTmp := &intermediate{}
  result, loaded := controller.meters.LoadOrStore(key, newTmp)
  newTmp.initOnce.Do(func() {. // where initOnce is a sync.Once
    result.inst = // ... initialize the underlying instrument once
  }
  // use result.inst
```

IOW like ea7bc599 :grin:

Comment (jmacd) [reply metric/registry/registry.go]: I was too hasty. Another commit is following. Also, this is more difficult to test.

File: sdk/metric/controller/basic/controller.go
Comment (jmacd) [reply sdk/metric/controller/basic/controller.go:348]: I removed the TODO and added support for multiple errors here: 3356eb5e

**Review by Aneurysm9 — APPROVED** [review]

File: metric/metrictest/meter.go
Comment (jmacd) [reply metric/metrictest/meter.go:160]: Hah, that wasn't doing anything, but based on your feedback I cleaned up all the locking in this file. The MeterProvider lock protects the batch of results, and I've added a couple of MeterProvider methods to avoid the `m.provider.lock` pattern.

File: metric/metrictest/meter.go
Comment (jmacd) [reply metric/metrictest/meter.go:172]: Fixed.

**Comment by jmacd** [comment]
This is ready to merge, I think. Thanks all for your patience.

**Comment by jmacd** [comment]
@MadVikingGod I reverted two commits that changed the controller to use a sync.Map behind its MeterProvider implementation. This can be addressed in a future PR. I've left a TODO about it, the spec will require this sort of improvement. I propose any further changes to this PR be left as TODOs, but I think it's ready to merge as-is.

**Comment by MadVikingGod** [comment]
I have talked with @jmacd offline about this, and agree that the sync.Map code is better off as an improvement in a later PR.  Count this as reapproving this from me.

**Comment by jmacd** [comment]
Thank you, approvers & maintainers. This was a tough one, I appreciate your patience. 🚀

**Pull request merged**
